server/routerlicious/packages/lambdas/src/scribe/checkpointManager.ts

```diff
@@ -136,19 +140,85 @@ export class CheckpointManager implements ICheckpointManager {
 		}
 	}
 
-	private async writeScribeCheckpointState(checkpoint: IScribe) {
-		await this.documentRepository.updateOne(
-			{
-				documentId: this.documentId,
-				tenantId: this.tenantId,
-			},
-			{
-				// MongoDB is particular about the format of stored JSON data. For this reason we store stringified
-				// given some data is user generated.
-				scribe: JSON.stringify(checkpoint),
-			},
-			null,
-		);
+	private async writeScribeCheckpointState(checkpoint: IScribe, noActiveClients: boolean) {
+		const lumberProperties = getLumberBaseProperties(this.documentId, this.tenantId);
+
+		if (this.localCheckpointEnabled) {
+			if (
+				this.localCheckpointCollection !== null &&
+				this.localCheckpointCollection !== undefined
+			) {
+				if (!noActiveClients) {
+					// If there are active clients, we write to the local collection
+					Lumberjack.info(`Writing checkpoint to local database`, lumberProperties);
+					// Use upsert in case document does not exist in the local database
+					await this.writeScribeCheckpointToCollection(checkpoint, true);
+				} else {
+					// No active clients, delete from local collection, write to global collection
+					Lumberjack.info(
+						`Removing checkpoint data from the local database. No active clients.`,
+						lumberProperties,
+					);
+					await Promise.all([
+						this.localCheckpointCollection
+							.deleteOne({
+								documentId: this.documentId,
+								tenantId: this.tenantId,
+							})
+							.catch((error) => {
+								Lumberjack.error(
+									`Error removing checkpoint data from the local database.`,
+									lumberProperties,
+									error,
+								);
+							}),
+						await this.writeScribeCheckpointToCollection(checkpoint, false),
+					]);
+				}
+			} else {
+				Lumberjack.info(
+					`No local collection found. Writing checkpoint to global database.`,
+					lumberProperties,
+				);
+				await this.writeScribeCheckpointToCollection(checkpoint, false);
+			}
+		} else {
+			Lumberjack.info(`Local checkpoints disabled. Writing checkpoint to global database.`);
+			await this.writeScribeCheckpointToCollection(checkpoint, false);
+		}
```

Comment:
The above code is too hard to maintain for the nested if else conditions. We can clearify logic by doing:
```
if (!this.localCheckpointEnabled){
  .....
 return ...
}

if (!this.localCheckpointCollection) {
 ....
 return ....
}

if (!noActiveClients) {
  ....
  return ....
}


.....

Reply:
Updated the logic